adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+        Returns
+        -------
+        member_ind : np.ndarray
+            Slice into this class's members.
+        """
+        # --- Integer Slice
+        # If the index is an integer then we need to convert it to a slice so that
+        # we do not change the dimensionality of the member arrays (or in the cases
+        # of a 1D array we need to avoid returning just a single value)
+        if isinstance(class_ind, int):
 
-        for k, v in other.__dict__.items():
-            self_v = self.__dict__[k]
+            ind = slice(class_ind, class_ind + 1)
 
-            if isinstance(v, np.ma.masked_array):
-                self.__dict__[k] = np.ma.concatenate([self_v, v])
-                self.__dict__[k].mask = np.concatenate([self_v.mask, v.mask])
-                self.__dict__[k].fill_value = self_v.fill_value
+        # --- Slices, Arrays, and Lists
+        elif isinstance(class_ind, (slice, np.ndarray, list)):
+            ind = class_ind
 
-            elif isinstance(v, np.ndarray):
-                self.__dict__[k] = np.concatenate([self_v, v])
+        else:
+            raise TypeError(
+                "class_ind should be one of {int, slice, np.ndarray, list}."
+            )
+
+        # --- Check boundaries on the slice
+        if isinstance(ind, slice) and ind.start is not None and ind.start >= len(self):
+            raise IndexError(f"Index {ind.start} is out of bounds.")
 
-            elif isinstance(v, Time):
-                self.__dict__[k] = Time(
-                    np.concatenate([self_v.mjd, v.mjd]), scale=v.scale, format="mjd"
+        elif isinstance(ind, slice) and self._class_index_to_members_is_slice:
+
+            # Check if the array of slices are consecutive and share
+            # the same step size. If so, create a single slice that
+            # combines all of the slices.
+            slices = self._class_index_to_members[ind]
+            is_consecutive = True
+            for i, s_i in enumerate(slices[:-1]):
+                if s_i.stop != slices[i + 1].start:
+                    is_consecutive = False
+                    break
+                if s_i.step is not None and (s_i.step != slices[i + 1].step):
+                    is_consecutive = False
+                    break
+
+            if is_consecutive:
+                logger.debug(
+                    "Slices are consecutive and share the same step. "
+                    f"Combining slices a single slice with start {slices[0].start}, "
+                    f"end {slices[-1].stop} and step {slices[0].step}."
                 )
+                return slice(slices[0].start, slices[-1].stop, slices[0].step)
+            else:
+                logger.debug(
+                    "Slices are not consecutive. "
+                    "Combining slices a concatenating the members index for each slice."
+                )
+                return np.concatenate([self._class_index[s] for s in slices])
```

Comment:
This is very tricky, subtle code. I'd recommend giving it a function and thorough unit tests. I'm not able to verify its correctness by reading.

The `break`s with a boolean status variable are a dead giveaway that the consecutivity check should be a separate function that returns True or False.

Reply:
Great point. I've extracted the consecutive slice check into a separate function `_check_slices_are_consecutive` which will return a boolean (added some unit tests too).